server/src/main/java/org/elasticsearch/cluster/ClusterState.java

```diff
@@ -840,4 +944,67 @@ public ClusterState apply(ClusterState state) {
 
     }
 
+    /**
+     * A collection of persistent node ids, denoting the voting configuration for cluster state changes.
+     */
+    public static class VotingConfiguration implements Writeable, ToXContentFragment {
+
+        public static final VotingConfiguration EMPTY_CONFIG = new VotingConfiguration(Collections.emptySet());
+
+        private final Set<String> nodeIds;
+
+        public VotingConfiguration(Set<String> nodeIds) {
+            this.nodeIds = Collections.unmodifiableSet(new HashSet<>(nodeIds));
+        }
+
+        public VotingConfiguration(StreamInput in) throws IOException {
+            nodeIds = Collections.unmodifiableSet(Sets.newHashSet(in.readStringArray()));
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeStringArray(nodeIds.toArray(new String[nodeIds.size()]));
+        }
+
+        public boolean hasQuorum(Collection<String> votes) {
+            final HashSet<String> intersection = new HashSet<>(nodeIds);
+            intersection.retainAll(votes);
+            return intersection.size() * 2 > nodeIds.size();
+        }
+
+        public Set<String> getNodeIds() {
+            return nodeIds;
+        }
+
+        @Override
+        public String toString() {
+            return "VotingConfiguration{" + String.join(",", nodeIds) + "}";
+        }
+
+        @Override
+        public boolean equals(Object o) {
```

Comment:
I don't see any assertions that this works - specifically that it ever returns `false`. Could we have some?